I actually see:

```
2024-09-05T07:07:29.3234814Z Caused by: io.quarkus.bootstrap.prebuild.CodeGenException: Failed to find com.google.protobuf:protoc:linux-x86_64:exe among dependencies
2024-09-05T07:07:29.3236628Z 	at io.quarkus.grpc.deployment.GrpcCodeGen.makeExecutableFromPath(GrpcCodeGen.java:447)
2024-09-05T07:07:29.3237932Z 	at io.quarkus.grpc.deployment.GrpcCodeGen.initExecutables(GrpcCodeGen.java:419)
2024-09-05T07:07:29.3239216Z 	at io.quarkus.grpc.deployment.GrpcCodeGen.trigger(GrpcCodeGen.java:144)
2024-09-05T07:07:29.3240252Z 	at io.quarkus.deployment.CodeGenerator.lambda$trigger$4(CodeGenerator.java:206)
2024-09-05T07:07:29.3241545Z 	at io.quarkus.deployment.CodeGenerator.callWithClassloader(CodeGenerator.java:181)
2024-09-05T07:07:29.3242792Z 	at io.quarkus.deployment.CodeGenerator.trigger(CodeGenerator.java:203)
2024-09-05T07:07:29.3243890Z 	at io.quarkus.deployment.CodeGenerator.initAndRun(CodeGenerator.java:80)
```

so I am still thinking this could be something else entirely...

--- a/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
@@ -522,13 +522,13 @@ public class TestResourceManager implements Closeable {
 
         // now we need to check whether the sets contain the exact same MATCHING_RESOURCE test resources
 
-        Set<TestResourceComparisonInfo> inExistingAndNotNext = onlyMatchingResourceItems(existing);
+        Set<TestResourceComparisonInfo> inExistingAndNotNext = new HashSet<>(existing);
         inExistingAndNotNext.removeAll(next);
         if (!inExistingAndNotNext.isEmpty()) {
             return true;
         }
 
-        Set<TestResourceComparisonInfo> inNextAndNotExisting = onlyMatchingResourceItems(next);
+        Set<TestResourceComparisonInfo> inNextAndNotExisting = new HashSet<>(next);
         inNextAndNotExisting.removeAll(existing);
         if (!inNextAndNotExisting.isEmpty()) {
             return true;
@@ -536,11 +536,6 @@ public class TestResourceManager implements Closeable {
 
         // the sets contain the same objects, so no need to reload
         return false;
-    }
-
-    private static Set<TestResourceComparisonInfo> onlyMatchingResourceItems(Set<TestResourceComparisonInfo> set) {
-        return set.stream().filter(i -> i.scope == MATCHING_RESOURCES).collect(
-                Collectors.toSet());
     }
 
     private static boolean hasRestrictedToClassScope(Set<TestResourceComparisonInfo> existing) {

--- a/test-framework/common/src/test/java/io/quarkus/test/common/TestResourceManagerReloadTest.java
+++ b/test-framework/common/src/test/java/io/quarkus/test/common/TestResourceManagerReloadTest.java
@@ -72,4 +72,16 @@ public class TestResourceManagerReloadTest {
                         new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
                         new TestResourceComparisonInfo("TEST", MATCHING_RESOURCES))));
     }
+
+    @Test
+    public void differentGlobalMultipleMatchingResource() {
+        assertTrue(testResourcesRequireReload(
+                Set.of(
+                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
+                        new TestResourceComparisonInfo("test4", GLOBAL)),
+                Set.of(new TestResourceComparisonInfo("test3", GLOBAL),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
+                        new TestResourceComparisonInfo("TEST", MATCHING_RESOURCES))));
+    }
 }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/TestResourceUtil.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/TestResourceUtil.java
@@ -43,9 +43,13 @@ final class TestResourceUtil {
             return Collections.emptySet();
         }
         Closeable closeable = state.testResourceManager;
-        if (closeable instanceof TestResourceManager testResourceManager) {
+        if (closeable == null) {
+            return Collections.emptySet();
+        }
+        // we can't compare with instanceof because of the different CL
+        if (TestResourceManager.class.getName().equals(closeable.getClass().getName())) {
             return TestResourceManagerReflections
-                    .testResourceComparisonInfo(testResourceManager);
+                    .testResourceComparisonInfo(closeable);
         }
         return Collections.emptySet();
     }

